File: cpp/MyCircularDeque.cpp
//
// Created by liyuan on 2020/3/29.
//
#include <iostream>
#include <deque>
#include <list>
using namespace std;

class MyCircularDeque {
private:
    list<int> contain;
    size_t max_size;
public:
    /** Initialize your data structure here. Set the size of the deque to be k. */
    MyCircularDeque(int k): max_size(k) {

    }

    /** Adds an item at the front of Deque. Return true if the operation is successful. */
    bool insertFront(int value) {
        if (contain.size() >= max_size) {
            return false;
        }
        contain.push_front(value);
        return true;
    }

    /** Adds an item at the rear of Deque. Return true if the operation is successful. */
    bool insertLast(int value) {
        if (contain.size() >= max_size) {
            return false;
        }
        contain.push_back(value);
        return true;
    }

    /** Deletes an item from the front of Deque. Return true if the operation is successful. */
    bool deleteFront() {
        if (contain.empty()) {
            return false;
        }
        contain.pop_front();
        return true;
    }

    /** Deletes an item from the rear of Deque. Return true if the operation is successful. */
    bool deleteLast() {
        if (contain.empty()) {
            return false;
        }
        contain.pop_back();
        return true;
    }

    /** Get the front item from the deque. */
    int getFront() {
        if (contain.empty()) {
            return -1;
        }
        return contain.front();
    }

    /** Get the last item from the deque. */
    int getRear() {
        if (contain.empty()) {
            return -1;
        }
        return contain.back();
    }

    /** Checks whether the circular deque is empty or not. */
    bool isEmpty() {
        return contain.empty();
    }

    /** Checks whether the circular deque is full or not. */
    bool isFull() {
        return contain.size() == max_size;
    }
};

int main() {
    /**
 * Your MyCircularDeque object will be instantiated and called as such:
     * */

    MyCircularDeque* obj = new MyCircularDeque(3);
    bool param_1 = obj->insertFront(1);
    bool param_2 = obj->insertLast(2);
    bool param_3 = obj->deleteFront();
    bool param_4 = obj->deleteLast();
    int param_5 = obj->getFront();
    int param_6 = obj->getRear();
    bool param_7 = obj->isEmpty();
    bool param_8 = obj->isFull();

}

File: python-pycharm/MyCircularDeque.py
#! -*- coding:utf-8 -*-

class Node(object):
    def __init__(self, value=-1):
        self.pre = None
        self.next = None
        self.value = value

class MyCircularDeque(object):
    def __init__(self, k):
        self._size = 0
        self._max_size = k
        self._node = Node()
        self._front = self._node
        self._last = self._node
        self._node.pre = self._node
        self._node.next = self._node

    def insert(self, position, value):
        node = Node(value)
        node.next = position
        position.pre.next = node
        node.pre = position.pre
        position.pre = node
        return node

    def insertFront(self, value):
        if self._size >= self._max_size:
            return False
        self.insert(self._front, value)
        self._front = self._last.next
        self._size += 1
        return True

    def insertLast(self, value):
        if self._size >= self._max_size:
            return False
        self.insert(self._last, value)
        self._front = self._last.next
        self._size += 1
        return True

    def deleteFront(self):
        if self._size < 1:
            return False
        self._last.next = self._front.next
        self._front.next.pre = self._last
        self._front = self._front.next
        self._size -= 1
        return True

    def deleteLast(self):
        if self._size < 1:
            return False
        last = self._last.pre
        last.pre.next = self._last
        self._last.pre = last.pre
        self._front = self._last.next
        self._size -= 1
        return True

    def getFront(self):
        if self._size < 1:
            return -1
        return self._front.value

    def getRear(self):
        if self._size < 1:
            return -1
        return self._last.pre.value

    def isEmpty(self):
        return self._size == 0

    def isFull(self):
        return self._size == self._max_size

if __name__ == '__main__':
    # Your MyCircularDeque object will be instantiated and called as such:
    obj = MyCircularDeque(3)
    param_1 = obj.insertFront(1)
    param_2 = obj.insertLast(2)
    param_3 = obj.deleteFront()
    param_4 = obj.deleteLast()
    param_5 = obj.getFront()
    param_6 = obj.getRear()
    param_7 = obj.isEmpty()
    param_8 = obj.isFull()